### DISCUSSION TITLE
Layer of prisms for gravity forward modelling

### DISCUSSION THREAD

**Discussion by santisoler** [pyvista/pyvista#1913] [Show and tell]
Hi! I'm Santiago Soler, one of the core developers of the [Fatiando a Terra](https://www.fatiando.org) project: open-source tools for geophysics. I'm leading the development of Harmonica, a library for processing and modelling gravity and magnetic data.

In order to obtain information about the structure of the solid Earth, we usually build models of buried bodies out of simple geometries. One of these geometries we use in Harmonica are rectangular prisms, that we use for example to approximate the topography of our areas of interested, and even the oceans down to the seafloor (bathymetry).
In Harmonica we represent a layer of prisms (a regular grid of rectangular prisms in the horizontal directions) whose top and bottom boundaries vary, draping a given surface, such as a topography or a bathymetry.
We have fancy ways of generating these type of models, although their visualization was very poor.

This week I started drafting a way to convert our layer of prisms into a `pyvista.UnstructuredGrid`, where each prism is represented by a `VTK_HEXAHEDRON`. You can take a look at the PR here: https://github.com/fatiando/harmonica/pull/291

Here is a simple example that read a topography grid for South Africa and shows how to build a prism layer, how to convert it to a `pyvista` grid and plot it:

https://github.com/fatiando/harmonica/blob/4eec55c14baef3870c2cfc08c6ebf555af759066/examples/visualization/prism_layer.py

The output looks like this:

![Screenshot_2021-12-03_16-11-52](https://user-images.githubusercontent.com/11541317/144659596-abc48ef5-d1bb-443e-b093-a577a1744a60.png)

> The colors represent the density contrast of the prisms. In this example we use only two density values: one for the relief and one for the ocean water.

We would really appreciate some feedback on it! The PR is still under development, so we are open to ideas to improve the implementation.

cc @leouieda

**Comment by banesullivan** [comment]
This is really awesome, @santisoler!! I'm stoked to see harmonica using PyVista!! It makes me so happy to see PyVista used more and more in the geophysics community!

If there is any interest in setting up PyVista with Sphinx-Gallery for [the examples in your documentation](https://www.fatiando.org/harmonica/latest/gallery/index.html), I would be happy to help get things up and running for you.

**Reply by banesullivan** [reply]
Scratch that... you're already using the SG image scraper! Fantastic!! https://github.com/fatiando/harmonica/pull/291/files#diff-e170e9a7d787c21095c6c11bb25f0f1ff0294a42a46d45ba6fb5ed794e457624R100

**Comment by santisoler** [comment]
Thanks @banesullivan! I'm glad you like it! I have to say that it was pretty straight forward to implement the conversion. Most of the thinks that need some work are more related to our own conventions within harmonica and Fatiando, but from a technical point of view, I'm amazed how easy it was to build the UnstructuredGrid.

Since you mention the docs: is there a way to add "interactive" plots in the Sphinx Gallery? I managed to make sphinx gallery to plot a static image out of the example (similar to the one I posted above), but I'm wondering if we can make it interactive. I saw that pyvista itself has some interactive plots in its docs, but I suspect they are not generated with sphinx gallery, right?
